pandas/core/dtypes/cast.py

```diff
@@ -253,72 +267,388 @@ def changeit():
+        elif is_datetime_or_timedelta_dtype(dtype):
+            # for datetime/timedelta, we need to return the underlying ints
+            fill_value = fill_array.astype(dtype)[0].value
         else:
-            dtype = np.object_
-            fill_value = np.nan
+            fill_value = fill_array.astype(dtype)[0]
+
+        return dtype, fill_value
     else:
-        dtype = np.object_
+        raise ValueError('fill_value must be a scalar, received '
+                         '{}'.format(type(fill_value)))
 
-    # in case we have a string that looked like a number
-    if is_extension_array_dtype(dtype):
-        pass
-    elif is_datetime64tz_dtype(dtype):
-        pass
-    elif issubclass(np.dtype(dtype).type, string_types):
-        dtype = np.object_
 
-    return dtype, fill_value
+def maybe_promote_with_array(dtype, fill_value=np.nan):
+    """
+    Determine minimal dtype to hold fill_value, when starting from dtype
+
+    This will also return the default missing value for the resulting dtype, if
+    necessary (e.g. for datetime / timedelta, the missing value will be `iNaT`)
+
+    Parameters
+    ----------
+    dtype : DType
+        The dtype to start from.
+    fill_value : np.ndarray / Series / Index
+        Array-like of values that the output dtype needs to be able to hold.
+
+    Returns
+    -------
+    dtype : DType
+        The updated dtype.
+    na_value : scalar
+        The missing value for the new dtype. Returns None or dtypes that
+        cannot hold missing values (integers, booleans, bytes).
+
+    See Also
+    --------
+    maybe_promote_with_scalar : similar method for scalar case
+
+    Examples
+    --------
+    >>> maybe_promote(np.dtype('int'), fill_value=np.array([None]))
+    (dtype('float64'), nan)
+    >>> maybe_promote(np.dtype('float'), fill_value=np.array(['abcd']))
+    (dtype('O'), nan)
+
+    For datetimes, timedeltas and datetimes with a timezone, the missing value
+    marker is pandas._libs.tslibs.iNaT (== np.iinfo('int64').min):
+
+    >>> maybe_promote(np.dtype('datetime64[ns]'),
+    ...               fill_value=np.array(['2018-01-01']))
+    (dtype('<M8[ns]'), -9223372036854775808)
+
+    The method will infer as conservatively as possible for integer types:
+
+    >>> maybe_promote(np.dtype('uint8'),
+    ...               fill_value=np.array([np.iinfo('uint8').max + 1]))
+    (dtype('uint16'), None)
+    >>> maybe_promote(np.dtype('uint8'), fill_value=np.array([-1]))
+    (dtype('int16'), None)
+    """
+
+    if isinstance(fill_value, np.ndarray):
+        if fill_value.ndim == 0:
+            # zero-dimensional arrays cannot be iterated over
+            fill_value = np.expand_dims(fill_value, 0)
+        elif fill_value.ndim > 1:
+            # ndarray, but too high-dimensional
+            fill_value = fill_value.ravel()
+    elif not isinstance(fill_value, (ABCSeries, ABCIndexClass)):
+        fill_type = type(fill_value)
```

Comment:
usually we use `type(foo).__name__`.  Any particular reason to not use the `.__name__` here?

Reply:
No, will adapt.